### PR TITLE
#873 873 changing map to vector in TabCHanger

### PR DESCRIPTION
**Pull request by e2l-su**
Changing std::map into a std::vector for performance gain over ModelTapChanger

### CODE REVIEW

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelPhaseTapChanger.cpp [hunk 1]
```diff
@@ -23,8 +23,8 @@
 
 namespace DYN {
 
-ModelPhaseTapChanger::ModelPhaseTapChanger(const std::string& id)
-    : ModelTapChanger(id),
+ModelPhaseTapChanger::ModelPhaseTapChanger(const std::string& id, int lowindex)
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelPhaseTapChanger.cpp:26]: Coding style: lowIndex

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelRatioTapChanger.cpp [hunk 2]
```diff
@@ -25,8 +25,9 @@
 namespace DYN {
 
 ModelRatioTapChanger::ModelRatioTapChanger(const std::string& id,
-                                           const std::string& side)
-    : ModelTapChanger(id),
+                                           const std::string& side,
+                                           int lowindex)
+    : ModelTapChanger(id, lowindex),
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelRatioTapChanger.cpp:30]: Same as above

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelTapChanger.h [hunk 3]
```diff
@@ -36,15 +36,16 @@ class ModelNetwork;
...
    * @param id : name of the tap changer
+   * @param lowindex : lowest step index
    */
-  inline explicit ModelTapChanger(const std::string& id)
+  inline explicit ModelTapChanger(const std::string& id, int lowindex)
       : id_(id),
         currentStepIndex_(0),
         regulating_(false),
-        lowStepIndex_(0),
+        lowStepIndex_(lowindex),
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelTapChanger.h:48]: Same as above

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelTapChanger.h [hunk 4]
```diff
@@ -66,20 +67,19 @@ class ModelTapChanger {
    * @return step associated to the index
    */
   inline const TapChangerStep& getStep(int key) const {
-    if (steps_.find(key) != steps_.end()) {
-      return steps_.find(key)->second;
-    } else {
+    try {
+      return steps_.at(key - getLowStepIndex());
+    } catch (...) {
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelTapChanger.h:72]: Generic catch should be limited, please put a specific error or a different way to catch this (if key >= steps_.size() for example)

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelTapChanger.h [hunk 5]
```diff
@@ -66,20 +67,19 @@ class ModelTapChanger {
    * @return step associated to the index
    */
   inline const TapChangerStep& getStep(int key) const {
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelTapChanger.h:69]: key->index

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelTapChanger.h [hunk 6]
```diff
@@ -172,7 +168,7 @@ class ModelTapChanger {
 
  private:
   std::string id_;  ///< id of the tap changer
-  std::map<int, TapChangerStep>
+  std::vector<TapChangerStep>
       steps_;             ///< map of TapChangerStep : index -> step
   int currentStepIndex_;  ///< index of the current step
   bool regulating_;       ///< is the tapChanger regulating ?
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelTapChanger.h:174]: Please add a const in front of lowStepIndex_

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelTwoWindingsTransformer.cpp [hunk 7]
```diff
@@ -26,65 +26,66 @@
...
-#include "DYNStepInterface.h"
 #include "DYNPhaseTapChangerInterface.h"
 #include "DYNRatioTapChangerInterface.h"
-#include "DYNBusInterface.h"
-#include "DYNModelConstants.h"
-#include "DYNModelNetwork.h"
-#include "DYNMessageTimeline.h"
-#include "DYNModelVoltageLevel.h"
+#include "DYNStepInterface.h"
+#include "DYNTrace.h"
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelTwoWindingsTransformer.cpp:52]: Please do not add a great quantity of styling related modifications (even if done automatically by a tool), it makes it very difficult to find where the modification were made and break the history when we want to compare to earlier versions. Please revert the modifications in this file not related to the update of the TapChanger API.

**Comment by e2l-su** [comment]
Hi,

That's ok to me, but how do I push ? I did not succeed pushing twice
yesterday.

One way could be to destroy the whole branch on serve side (?)

See you :)



On Mon, 27 Jul 2020 23:28:26 -0700
rosiereflo <notifications@github.com> wrote:

> @rosiereflo requested changes on this pull request.
> 
> Looks globally good to me, all comments are related to coding style.
> 
> > @@ -23,8 +23,8 @@
>  
>  namespace DYN {
>  
> -ModelPhaseTapChanger::ModelPhaseTapChanger(const std::string& id)
> -    : ModelTapChanger(id),
> +ModelPhaseTapChanger::ModelPhaseTapChanger(const std::string& id,
> int lowindex)
> 
> Coding style: lowIndex 
> 
> > @@ -25,8 +25,9 @@
>  namespace DYN {
>  
>  ModelRatioTapChanger::ModelRatioTapChanger(const std::string& id,
> -                                           const std::string& side)
> -    : ModelTapChanger(id),
> +                                           const std::string& side,
> +                                           int lowindex)
> +    : ModelTapChanger(id, lowindex),
> 
> Same as above
> 
> >        : id_(id),
>          currentStepIndex_(0),
>          regulating_(false),
> -        lowStepIndex_(0),
> +        lowStepIndex_(lowindex),
> 
> Same as above
> 
> > @@ -66,20 +67,19 @@ class ModelTapChanger {
>     * @return step associated to the index
>     */
>    inline const TapChangerStep& getStep(int key) const {
> -    if (steps_.find(key) != steps_.end()) {
> -      return steps_.find(key)->second;
> -    } else {
> +    try {
> +      return steps_.at(key - getLowStepIndex());
> +    } catch (...) {
> 
> Generic catch should be limited, please put a specific error or a
> different way to catch this (if key >= steps_.size() for example)
> 
> > @@ -66,20 +67,19 @@ class ModelTapChanger {
>     * @return step associated to the index
>     */
>    inline const TapChangerStep& getStep(int key) const {
> 
> key->index
> 
> > @@ -172,7 +168,7 @@ class ModelTapChanger {
>  
>   private:
>    std::string id_;  ///< id of the tap changer
> -  std::map<int, TapChangerStep>
> +  std::vector<TapChangerStep>
>        steps_;             ///< map of TapChangerStep : index -> step
>    int currentStepIndex_;  ///< index of the current step
>    bool regulating_;       ///< is the tapChanger regulating ?
> 
> Please add a const in front of lowStepIndex_
> 
> >  #include "DYNPhaseTapChangerInterface.h"
>  #include "DYNRatioTapChangerInterface.h"
> -#include "DYNBusInterface.h"
> -#include "DYNModelConstants.h"
> -#include "DYNModelNetwork.h"
> -#include "DYNMessageTimeline.h"
> -#include "DYNModelVoltageLevel.h"
> +#include "DYNStepInterface.h"
> +#include "DYNTrace.h"
> 
> Please do not add a great quantity of styling related modifications
> (even if done automatically by a tool), it makes it very difficult to
> find where the modification were made and break the history when we
> want to compare to earlier versions. Please revert the modifications
> in this file not related to the update of the TapChanger API.
> 



-------------------------

Cordialement, Denis Gillain, http://www.ellipsys.fr
__________________________________________________

E      L      L      I      P      S      Y      S

+33 (0) 1 42 36 37 19 -  denis.gillain@ellipsys.fr
__________________________________________________

File: dynawo/sources/Models/CPP/ModelNetwork/test/TestTapChanger.cpp [hunk 8]
```diff
@@ -68,27 +66,27 @@ TEST(ModelsModelNetwork, ModelNetworkTapChanger) {
...
 
-  ptc.addStep(1, TapChangerStep(1., 1., 1., 1., 1., 1.));
-  ptc.addStep(10, TapChangerStep(10., 10., 10., 10., 10., 10.));
-  ptc.addStep(2, TapChangerStep(2., 2., 2., 2., 2., 2.));
+  ptc.addStep(TapChangerStep(1., 1., 1., 1., 1., 1.));
+  ptc.addStep(TapChangerStep(10., 10., 10., 10., 10., 10.));
+  ptc.addStep(TapChangerStep(2., 2., 2., 2., 2., 2.));
 
-  const TapChangerStep& tcs = ptc.getCurrentStep();
+  const TapChangerStep& tcs = ptc.getStep(2);
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/ModelNetwork/test/TestTapChanger.cpp:73]: Please leave the test of getCurrentStep()

**Comment by e2l-su** [comment]
I rollback the pull request since changing workstation. I will pull another request later to do the job.

**Pull request closed without merging**